examples/java/src/main/java/org/apache/beam/examples/io/examplekafkaread/ExampleKafkaReadIO.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.dataflow.qual.Pure;
+
+/**
+ * An example unbounded SDF IO, using Kafka as an underlying source. This IO deliberately has a
+ * minimal set of Kafka features, and exists as an example of how to write an unbounded SDF source.
+ *
+ * <p>This IO should not be used in pipelines, and has no guarantees for quality, correctness, or
+ * performance.
+ *
+ * <p>This IO was generated as a pared down version of KafkaIO, and should act to consume
+ * KV<byte[],byte[]> pairs from Kafka In practice, this IO would be used thusly: pipeline
+ * .apply(ExampleKafkaReadIO.read() .withBootstrapServers("broker_1:9092,broker_2:9092")
+ * .withTopics(ImmutableList.of("my_topic"))) .apply(Some Other Transforms);
+ */
+public class ExampleKafkaReadIO {
+
+  public static Read read() {
+    return new AutoValue_ExampleKafkaReadIO_Read.Builder()
+        .setTopics(new ArrayList<>())
+        .setConsumerConfig(ExampleKafkaReadIOUtils.DEFAULT_CONSUMER_PROPERTIES)
+        .build();
+  }
+
+  @AutoValue
+  @AutoValue.CopyAnnotations
+  public abstract static class Read extends PTransform<PBegin, PCollection<KV<byte[], byte[]>>> {
+
+    @Pure
+    abstract Map<String, Object> getConsumerConfig();
+
+    @Pure
+    abstract List<String> getTopics();
+
+    abstract Read.Builder toBuilder();
+
+    @Experimental(Kind.PORTABILITY)
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConsumerConfig(Map<String, Object> config);
+
+      abstract Builder setTopics(List<String> topics);
+
+      abstract Read build();
+    }
+
+    /**
+     * This expand method is characteristic of many top level expands for IOs. It validates required
+     * parameters and configurations, and then it applies a series of DoFns to actually execute the
+     * read or write. It is typical to have multiple DoFns here, as frequently an IO needs to
+     * determine what to read from the configuration before it actually reads. In this case, we are
+     * provided a topic name, but Kafka further divides its topics into partitions. As such, {@link
+     * GenerateTopicPartitions} exists to query Kafka for this detail, yielding one or more {@link
+     * org.apache.kafka.common.TopicPartition}s for every topic name.
+     */
+    @Override
+    public PCollection<KV<byte[], byte[]>> expand(@NonNull PBegin input) {
+      checkArgument(
+          getConsumerConfig().get(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG) != null,
+          "withBootstrapServers() is required");
+      checkArgument(getTopics() != null && getTopics().size() > 0, "withTopics() is required");
+      return input
+          .apply(Impulse.create())
+          .apply(ParDo.of(new GenerateTopicPartitions(getConsumerConfig(), getTopics())))
```

Comment:
There should be a Reshuffle between these two. Otherwise topic partition generation and reading gets fused and everything will happen in a single worker.

Also, a failure when reading will result in topic-partitions being regenerated.